lib/net/ipnet.go

```diff
@@ -31,33 +31,60 @@ func (i IPNet) MarshalJSON() ([]byte, error) {
 
 // UnmarshalJSON interface for an IPNet
 func (i *IPNet) UnmarshalJSON(b []byte) error {
+	// The YAML format will be a string, so unmarshal into a string.
 	var s string
 	if err := json.Unmarshal(b, &s); err != nil {
 		return err
 	}
-	if _, ipnet, err := net.ParseCIDR(s); err != nil {
-		return err
-	} else {
-		i.IP = ipnet.IP
+
+	// First attempt to parse as a CIDR and then as an IP address (assuming
+	// full mask).  When parsing as a CIDR we preserve the actual IP address
+	// and *not* the masked IP.
+	if ip, ipnet, err := ParseCIDR(s); err == nil {
+		i.IP = ip.IP
 		i.Mask = ipnet.Mask
 		return nil
+	} else if err = i.IP.UnmarshalText([]byte(s)); err == nil {
+		ip := &IP{i.IP}
+		n := ip.Network()
+		i.IPNet = n.IPNet
+		return nil
+	} else {
+		return err
 	}
 }
 
-// Version returns the IP version for an IPNet
+// Version returns the IP version for an IPNet, or 0 if not a valid IP net.
 func (i *IPNet) Version() int {
-	if i.IP.To4() == nil {
+	if i.IP.To4() != nil {
+		return 4
+	} else if len(i.IP) == net.IPv6len {
 		return 6
 	}
-	return 4
+	return 0
+}
+
+// Network returns the masked IP network.
+func (i *IPNet) Network() *IPNet {
+	_, n, _ := ParseCIDR(i.String())
+	return n
 }
 
 func ParseCIDR(c string) (*IP, *IPNet, error) {
 	netIP, netIPNet, e := net.ParseCIDR(c)
-	if netIPNet == nil {
+	if netIPNet == nil || e != nil{
 		return nil, nil, e
 	}
-	return &IP{netIP}, &IPNet{*netIPNet}, e
+	ip := &IP{netIP}
+	ipnet := &IPNet{*netIPNet}
+
+	// If the IP version is IPv4, normalise it to 4-byte format so that it
+	// is in the same format as the IPNet.
+	if ip4 := ip.IP.To4(); ip4 != nil {
```

Comment:
I'll update  the comment.

basically it makes it easier to flip between IPs and the IP within an IPNet if they use the same length encoding (it's primarily a testing and debug thing).